File: sdks/local-rp/php/tests/BeginTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/bootstrap.php';

use LinkKeys\LocalRp\Begin;
use LinkKeys\LocalRp\BeginLocalLoginConfig;
use LinkKeys\LocalRp\GenerateLocalRpIdentityConfig;
use LinkKeys\LocalRp\Identity;
use LinkKeys\LocalRp\PendingLogin;

function beginTestMaterial(): \LinkKeys\LocalRp\LocalRpKeyMaterial
{
    return Identity::generateLocalRpIdentity(new GenerateLocalRpIdentityConfig('Test App', new \DateTimeImmutable('now')));
}

TestKit::test('begin.defaults_claims_and_produces_pending_state', function () {
    $m = beginTestMaterial();
    [$redirect, $pending] = Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'http://localhost:8080/callback', 'example.com', new \DateTimeImmutable('now')));

    TestKit::assertTrue(str_starts_with($redirect->redirectUrl, 'https://example.com/auth/local-rp?signed_request='));
    TestKit::assertEquals('example.com', $pending->userDomain);
    TestKit::assertEquals('http://localhost:8080/callback', $pending->callbackUrl);
    TestKit::assertEquals(32, strlen($pending->nonce));
    TestKit::assertEquals(32, strlen($pending->state));
});

TestKit::test('begin.rejects_non_http_callback_scheme', function () {
    $m = beginTestMaterial();
    TestKit::assertThrows(fn () => Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'myapp://callback', 'example.com', new \DateTimeImmutable('now'))));
});

TestKit::test('begin.rejects_empty_user_domain', function () {
    $m = beginTestMaterial();
    TestKit::assertThrows(fn () => Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'http://localhost/callback', '', new \DateTimeImmutable('now'))));
});

TestKit::test('begin.two_calls_never_reuse_nonce_or_state', function () {
    $m = beginTestMaterial();
    [, $p1] = Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'http://localhost/callback', 'example.com', new \DateTimeImmutable('now')));
    [, $p2] = Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'http://localhost/callback', 'example.com', new \DateTimeImmutable('now')));
    TestKit::assertTrue($p1->nonce !== $p2->nonce);
    TestKit::assertTrue($p1->state !== $p2->state);
});

TestKit::test('begin.pending_login_array_round_trip', function () {
    $m = beginTestMaterial();
    [, $pending] = Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'http://localhost/callback', 'example.com', new \DateTimeImmutable('now')));
    $roundTripped = PendingLogin::fromArray($pending->toArray());
    TestKit::assertEquals($pending->nonce, $roundTripped->nonce);
    TestKit::assertEquals($pending->state, $roundTripped->state);
    TestKit::assertEquals($pending->userDomain, $roundTripped->userDomain);
    TestKit::assertEquals($pending->callbackUrl, $roundTripped->callbackUrl);
    TestKit::assertEquals($pending->requiredClaims, $roundTripped->requiredClaims);
});

TestKit::test('begin.pending_login_retains_required_claims', function () {
    $m = beginTestMaterial();
    [, $pending] = Begin::beginLocalLogin(new BeginLocalLoginConfig(
        $m,
        'http://localhost/callback',
        'example.com',
        new \DateTimeImmutable('now'),
        null,
        ['handle', 'email']
    ));
    TestKit::assertEquals(['handle', 'email'], $pending->requiredClaims);

    [, $defaultPending] = Begin::beginLocalLogin(new BeginLocalLoginConfig($m, 'http://localhost/callback', 'example.com', new \DateTimeImmutable('now')));
    TestKit::assertEquals(Begin::DEFAULT_REQUIRED_CLAIMS, $defaultPending->requiredClaims);
});

TestKit::test('begin.default_and_custom_claims', function () {
    $m = beginTestMaterial();
    TestKit::assertEquals(['display_name', 'email', 'handle'], Begin::DEFAULT_REQUESTED_CLAIMS);
    TestKit::assertEquals(['handle'], Begin::DEFAULT_REQUIRED_CLAIMS);
});

exit(TestKit::summary('BeginTest') === 0 ? 0 : 1);

File: sdks/local-rp/dart/test/begin_test.dart
import 'package:linkkeys_local_rp/linkkeys_local_rp.dart';
import 'package:test/test.dart';

Future<LocalRpKeyMaterial> _material() =>
    generateLocalRpIdentity(GenerateLocalRpIdentityConfig(
        appName: 'Test App', now: DateTime.now().toUtc()));

void main() {
  group('Begin', () {
    test('beginLocalLogin defaults claims and produces pending state',
        () async {
      final m = await _material();
      final result = await beginLocalLogin(BeginLocalLoginConfig(
        keyMaterial: m,
        callbackUrl: 'http://localhost:8080/callback',
        userDomain: 'example.com',
        now: DateTime.now().toUtc(),
      ));

      expect(result.redirect.redirectUrl,
          startsWith('https://example.com/auth/local-rp?signed_request='));
      expect(result.pending.userDomain, equals('example.com'));
      expect(
          result.pending.callbackUrl, equals('http://localhost:8080/callback'));
      expect(result.pending.nonce.length, equals(32));
      expect(result.pending.state.length, equals(32));
      expect(result.pending.requiredClaims, equals(defaultRequiredClaims));
    });

    // FIX A.1: `PendingLogin` must retain `requiredClaims` so
    // `completeLocalLogin` can enforce it later -- an app that persists and
    // reloads `PendingLogin` between `begin` and `complete` (its documented
    // responsibility) must get back exactly what the login was begun with,
    // not silently fall back to the defaults.
    test('beginLocalLogin retains caller-supplied required claims in pending',
        () async {
      final m = await _material();
      final result = await beginLocalLogin(BeginLocalLoginConfig(
        keyMaterial: m,
        callbackUrl: 'http://localhost/callback',
        userDomain: 'example.com',
        requiredClaims: const ['email', 'handle'],
        now: DateTime.now().toUtc(),
      ));
      expect(result.pending.requiredClaims, equals(['email', 'handle']));

      // Simulate the app persisting-and-reloading PendingLogin (its
      // documented responsibility -- this SDK owns no storage): rebuilding
      // it from the exact field values must round-trip requiredClaims.
      final reloaded = PendingLogin(
        nonce: result.pending.nonce,
        state: result.pending.state,
        userDomain: result.pending.userDomain,
        callbackUrl: result.pending.callbackUrl,
        requiredClaims: result.pending.requiredClaims,
      );
      expect(reloaded.requiredClaims, equals(result.pending.requiredClaims));
    });

    test('beginLocalLogin accepts empty required claims', () async {
      final m = await _material();
      final result = await beginLocalLogin(BeginLocalLoginConfig(
        keyMaterial: m,
        callbackUrl: 'http://localhost/callback',
        userDomain: 'example.com',
        requiredClaims: const [],
        now: DateTime.now().toUtc(),
      ));
      expect(result.pending.requiredClaims, isEmpty);
    });

    test('beginLocalLogin rejects non-http callback scheme', () async {
      final m = await _material();
      await expectLater(
        beginLocalLogin(BeginLocalLoginConfig(
          keyMaterial: m,
          callbackUrl: 'myapp://callback',
          userDomain: 'example.com',
          now: DateTime.now().toUtc(),
        )),
        throwsA(isA<SdkException>()),
      );
    });

    test('beginLocalLogin rejects empty user domain', () async {
      final m = await _material();
      await expectLater(
        beginLocalLogin(BeginLocalLoginConfig(
          keyMaterial: m,
          callbackUrl: 'http://localhost/callback',
          userDomain: '',
          now: DateTime.now().toUtc(),
        )),
        throwsA(isA<SdkException>()),
      );
    });

    test('beginLocalLogin two calls never reuse nonce or state', () async {
      final m = await _material();
      final r1 = await beginLocalLogin(BeginLocalLoginConfig(
        keyMaterial: m,
        callbackUrl: 'http://localhost/callback',
        userDomain: 'example.com',
        now: DateTime.now().toUtc(),
      ));
      final r2 = await beginLocalLogin(BeginLocalLoginConfig(
        keyMaterial: m,
        callbackUrl: 'http://localhost/callback',
        userDomain: 'example.com',
        now: DateTime.now().toUtc(),
      ));
      expect(r1.pending.nonce, isNot(equals(r2.pending.nonce)));
      expect(r1.pending.state, isNot(equals(r2.pending.state)));
    });
  });
}
